syntax = "proto2";

package rv.protos;

import "rastervision/protos/machine_learning.proto";
import "rastervision/protos/scene.proto";
import "rastervision/protos/make_training_chips.proto";
import "rastervision/protos/train.proto";
import "rastervision/protos/predict.proto";
import "rastervision/protos/eval.proto";
import "rastervision/protos/raster_transformer.proto";
import "rastervision/protos/label_store.proto";

message ChainWorkflowConfig {
    /*
        train_scenes are used by the make_training_chips and train commands to
        generate chips to train the model. Each scene must contain a
        ground_truth_label_store.
    */
    repeated Scene train_scenes = 1;

    /*
        validation_scenes are used by the make_training_chips, train, predict,
        and eval commands to generate chips to evaluate the model. Each scene
        must contain a ground_truth_label_store. The prediction_label_stores
        should not be included in these scenes as they will be generated
        automatically.
    */
    repeated Scene validation_scenes = 18;

    /*
        test_scenes are an optional additional set of scenes used to evaluate
        the model after training has taken place. If they contain
        ground_truth_label_stores, they are used by the predict and eval
        commands. If they don't, they are only used by the predict command. The
        prediction_label_stores should not be included in these scenes as they
        will be generated automatically.
    */
    repeated Scene test_scenes = 20;

    /*
        The label_store_template is used to fill in any missing LabelStore
        fields and is also used to generate prediction_label_stores
        automatically for validation and test_scenes.
    */
    required LabelStore label_store_template = 21;

    required MachineLearning machine_learning = 3;
    required RasterTransformer raster_transformer = 5;
    required int32 chip_size = 6;
    required bool debug = 7;

    // Map of parameters to values to use when doing parameter substitution on
    // generated URIs. One of the keys must be 'rv_root' which is used when
    // generating the URI for all the output generated by running RV.
    map<string, string> local_uri_map = 8;
    map<string, string> remote_uri_map = 9;

    // The names of the raw dataset, (processed) dataset, model, prediction,
    // and eval. These are used to generate URIs.
    required string raw_dataset_key = 19;
    required string dataset_key = 10;
    required string model_key = 11;
    required string prediction_key = 12;
    required string eval_key = 13;

    required MakeTrainingChipsConfig.Options make_training_chips_options = 14;
    required TrainConfig.Options train_options = 15;
    required PredictConfig.Options predict_options = 16;
    required EvalConfig.Options eval_options = 17;
}
